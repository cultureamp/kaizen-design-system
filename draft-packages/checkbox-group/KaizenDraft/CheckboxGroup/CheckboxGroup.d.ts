import * as React from "react";
export declare type CheckboxGroupProps = {
    automationId?: string;
    labelText: string | React.ReactNode;
    noBottomMargin?: boolean;
};
declare type CheckboxGroup = React.FunctionComponent<CheckboxGroupProps>;
declare const CheckboxGroup: CheckboxGroup;
export default CheckboxGroup;
