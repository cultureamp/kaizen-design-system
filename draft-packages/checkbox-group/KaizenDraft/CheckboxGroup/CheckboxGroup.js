"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var classnames_1 = __importDefault(require("classnames"));
var React = __importStar(require("react"));
var draft_1 = require("@kaizen/component-library/draft");
var styles = require("./styles.scss");
var CheckboxGroup = function (_a) {
    var _b;
    var automationId = _a.automationId, children = _a.children, labelText = _a.labelText, _c = _a.noBottomMargin, noBottomMargin = _c === void 0 ? false : _c;
    return (React.createElement("div", { "data-automation-id": automationId && automationId + "-field-checkbox-group", className: classnames_1["default"](styles.checkboxGroupContainer, (_b = {},
            _b[styles.noBottomMargin] = noBottomMargin,
            _b)) },
        React.createElement("div", { className: styles.checkboxGroupLabel },
            React.createElement(draft_1.Label, { automationId: automationId + "-field-label", labelText: labelText, labelType: "text" })),
        children));
};
exports["default"] = CheckboxGroup;
//# sourceMappingURL=CheckboxGroup.js.map