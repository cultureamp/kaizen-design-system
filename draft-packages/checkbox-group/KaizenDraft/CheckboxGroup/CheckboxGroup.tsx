import { Label } from "@kaizen/draft-form"
import classnames from "classnames"
import * as React from "react"

const styles = require("./styles.scss")

export type CheckboxGroupProps = {
  automationId?: string
  labelText: string | React.ReactNode
  noBottomMargin?: boolean
}

type CheckboxGroup = React.FunctionComponent<CheckboxGroupProps>

const CheckboxGroup: CheckboxGroup = ({
  automationId,
  children,
  labelText,
  noBottomMargin = false,
}) => (
  <div
    data-automation-id={automationId && `${automationId}-field-checkbox-group`}
    className={classnames(styles.checkboxGroupContainer, {
      [styles.noBottomMargin]: noBottomMargin,
    })}
  >
    <div className={styles.checkboxGroupLabel}>
      <Label
        automationId={`${automationId}-field-label`}
        labelText={labelText}
        labelType="text"
      />
    </div>
    {children}
  </div>
)

export default CheckboxGroup
