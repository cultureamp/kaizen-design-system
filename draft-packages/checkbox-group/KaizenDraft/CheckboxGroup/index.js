"use strict";
exports.__esModule = true;
var CheckboxGroup_1 = require("./CheckboxGroup");
exports.CheckboxGroup = CheckboxGroup_1["default"];
//# sourceMappingURL=index.js.map