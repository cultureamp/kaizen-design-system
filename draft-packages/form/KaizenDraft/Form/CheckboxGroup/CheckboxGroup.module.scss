@import "~@kaizen/design-tokens/sass/spacing";

.checkboxGroupLabel {
  padding-bottom: $spacing-xs;
}
.checkboxGroupContainer {
  display: flex;
  flex-direction: column;
  gap: $spacing-xs;
}
