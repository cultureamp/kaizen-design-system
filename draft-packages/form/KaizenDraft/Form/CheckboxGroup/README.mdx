---
title: Checkbox Group
navTitle: Checkbox Group
summaryParagraph: A checkbox group contains checkbox fields that let people select zero or more options from a set.
tags: ["Choice list", "Radio button", "Option button", "Choice", "Selection"]
githubLabels: ["component:checkbox", "component:form"]
needToKnow:
- Prefer vertical (stacked) checkboxes to easily scan and compare options.
- When choices are mutually exclusive, use a radio instead.
- For settings that are immediately applied, consider using a toggle switch instead.
demoStoryId: components-form-checkbox-group--interactive-kaizen-site-demo
headerImage: checkbox-group
health:
  designed: true
  documented: true
  implemented: true
  latestDesign: false
  allVariants: false
  responsive: true
  internationalized: true
  accessible: true
---

import WhenToUseAndWhenNotToUse from "docs-components/WhenToUseAndWhenNotToUse"
import WhenToUse from "docs-components/WhenToUse"
import WhenNotToUse from "docs-components/WhenNotToUse"

## Visuals

<iframe style="border: none;" width="744" height="450" src="https://www.figma.com/embed?embed_host=share&url=https%3A%2F%2Fwww.figma.com%2Ffile%2FeZKEE5kXbEMY3lx84oz8iN%2F%25E2%259D%25A4%25EF%25B8%258F-UI-Kit-Heart%3Fnode-id%3D3867%253A0" allowfullscreen></iframe>

## To keep in mind

*   Show options in a sensible order, such as alphabetical, numeric.
*   Avoid long lists. Chunk long lists and group related options together.
*   The setting only applies when the user takes an action, such as clicking a save button. The rare exception is for filtering checkboxes, which apply instantly.
*   Use specific, clear, concrete labels.
*   Prefer vertical (stacked) checkbox fields to easily scan and compare options.
*   Offer a default option.
*   Avoid nesting.
*   Disabled state
    *   See [Interaction states guidelines for disabled elements](/guidelines/interaction-states#disabled-state).
*   Focus indicator
    *   See [Interaction states guidelines for focus indicators](/guidelines/interaction-states#focus-indicator).
       *   On keyboard navigation and mouse click, show the focus indicator on the checkbox only. 
       *   Although the label is still navigable/clickable when selecting a checkbox, exclude showing the focus indicator on the label.

## When to use and when not to use

<WhenToUseAndWhenNotToUse>
<WhenToUse>

*   Single choice options
*   Shown side-by-side or stacked
*   Where there is space to show them all at once

</WhenToUse>
<WhenNotToUse>

*   Donâ€™t use when there is more than 20 options.
*   When choices are mutually exclusive, use a [radio](/components/radio) instead.
*   For settings that are immediately applied, consider using a [toggle switch](/components/toggle-switch-field) instead.

</WhenNotToUse>
</WhenToUseAndWhenNotToUse>

## See also

- [Toggle switch](/components/toggle-switch-field)
- [Radio Group](/components/radio-group)

## External links

Here are some examples of other existing design systems:

- [Material: form fields](https://material.io/develop/web/components/input-controls/form-fields/)
- [Polaris: choice list](https://polaris.shopify.com/components/forms/choice-list)
