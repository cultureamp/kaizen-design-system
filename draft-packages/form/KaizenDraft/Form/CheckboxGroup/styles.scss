@import "~@kaizen/deprecated-component-library-helpers/styles/color";
@import "~@kaizen/deprecated-component-library-helpers/styles/type";

.checkboxGroupLabel {
  margin-bottom: $ca-grid / 2;
}
.checkboxGroupContainer {
  display: flex;
  flex-direction: column;
  margin-bottom: $ca-grid / 2;
  &.noBottomMargin {
    margin-bottom: 0;
  }

  label {
    width: 400px;
  }
}
