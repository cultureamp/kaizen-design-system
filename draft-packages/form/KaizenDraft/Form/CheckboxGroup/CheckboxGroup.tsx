import React, { useState, HTMLAttributes } from "react"
import classnames from "classnames"
import { v4 } from "uuid"
import { OverrideClassName } from "@kaizen/component-base"
import { Label } from "../Primitives"
import styles from "./CheckboxGroup.module.scss"

export interface CheckboxGroupProps
  extends OverrideClassName<HTMLAttributes<HTMLDivElement>> {
  children?: React.ReactNode
  labelText: string | React.ReactNode
  labelId?: string
  noBottomMargin?: boolean
  reversed?: boolean
  /**
   * **Deprecated:** Use test id compatible with your testing library (eg. `data-testid`).
   * @deprecated
   */
  automationId?: string
}

/**
 * {@link https://cultureamp.design/components/checkbox-group/ Guidance} |
 * {@link https://cultureamp.design/storybook/?path=/docs/components-form-checkbox-group--interactive-kaizen-site-demo Storybook}
 */
export const CheckboxGroup = ({
  children,
  labelText,
  labelId: propsLabelId,
  noBottomMargin = false,
  reversed = false,
  automationId,
  classNameOverride,
  ...restProps
}: CheckboxGroupProps): JSX.Element => {
  const [labelId] = useState<string>(propsLabelId || v4())
  return (
    <div
      data-automation-id={
        automationId && `${automationId}-field-checkbox-group`
      }
      className={classnames(
        styles.checkboxGroupContainer,
        classNameOverride,
        noBottomMargin && styles.noBottomMargin,
        reversed && styles.reversed
      )}
      role="group"
      aria-labelledby={labelId}
      {...restProps}
    >
      <div className={styles.checkboxGroupLabel}>
        <Label
          automationId={automationId && `${automationId}-field-label`}
          id={labelId}
          labelText={labelText}
          labelType="text"
          reversed={reversed}
        />
      </div>
      {children}
    </div>
  )
}

CheckboxGroup.displayName = "CheckboxGroup"
