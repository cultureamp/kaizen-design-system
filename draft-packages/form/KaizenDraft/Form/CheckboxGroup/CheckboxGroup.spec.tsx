import React from "react"
import { render, screen } from "@testing-library/react"
import { CheckboxGroup, CheckboxGroupProps } from "./CheckboxGroup"

const defaultCheckboxGroupProps: CheckboxGroupProps = {
  automationId: "CheckboxGroupAutomationId",
  labelText: "Label",
  noBottomMargin: true,
}
const renderCheckboxGroupProps = (
  props?: CheckboxGroupProps
): ReturnType<typeof render> => {
  const mergedCheckboxGroupProps = { ...defaultCheckboxGroupProps, ...props }

  return render(<CheckboxGroup {...mergedCheckboxGroupProps} />)
}

describe("<CheckboxGroup />", () => {
  it("renders a `data-automation-id` attribute", () => {
    const { container } = renderCheckboxGroupProps()
    expect(
      container.querySelector(
        `[data-automation-id="${defaultCheckboxGroupProps.automationId}-field-checkbox-group"]`
      )
    ).toBeTruthy()
  })

  it("renders a title", () => {
    const title = "Label"

    const { queryByText } = render(<CheckboxGroup labelText={title} />)
    expect(queryByText(title)).toBeTruthy()
  })

  describe("accessibilty", () => {
    it("has an accessible name when not provided a labelId", () => {
      render(<CheckboxGroup labelText="Label title" />)
      expect(
        screen.getByRole("group", { name: "Label title" })
      ).toBeInTheDocument()
    })
    it("has an accessible name when provided a labelId", () => {
      render(<CheckboxGroup labelText="Label title" labelId="some-custom-id" />)
      expect(
        screen.getByRole("group", { name: "Label title" })
      ).toBeInTheDocument()
    })
  })
})
