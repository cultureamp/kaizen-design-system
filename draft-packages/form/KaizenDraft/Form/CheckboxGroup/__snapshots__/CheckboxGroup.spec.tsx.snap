// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxGroup />  renders Checkbox Group with noBottomMargin attribute 1`] = `
<div
  aria-labelledby="81432aa4-255d-4e2f-9b4d-a0b0ea11acd9"
  class="checkboxGroupContainer noBottomMargin"
  data-automation-id="CheckboxGroupAutomationId-field-checkbox-group"
  role="group"
>
  <div
    class="checkboxGroupLabel"
  >
    <label
      class="label text"
      data-automation-id="CheckboxGroupAutomationId-field-label"
      id="81432aa4-255d-4e2f-9b4d-a0b0ea11acd9"
    >
      <span
        class="appendedLabel"
      >
        CheckboxGroup Label
      </span>
    </label>
  </div>
</div>
`;

exports[`<CheckboxGroup />  renders Checkbox Group without noBottomMargin attribute 1`] = `
<div
  aria-labelledby="0c14ab49-d68b-4074-b48a-14db60c75c87"
  class="checkboxGroupContainer"
  data-automation-id="CheckboxGroupAutomationId-field-checkbox-group"
  role="group"
>
  <div
    class="checkboxGroupLabel"
  >
    <label
      class="label text"
      data-automation-id="CheckboxGroupAutomationId-field-label"
      id="0c14ab49-d68b-4074-b48a-14db60c75c87"
    >
      <span
        class="appendedLabel"
      >
        CheckboxGroup Label
      </span>
    </label>
  </div>
</div>
`;
